javalib/src/main/scala/java/util/ArrayDeque.scala

```diff
@@ -0,0 +1,241 @@
+/*
+ * Scala.js (https://www.scala-js.org/)
+ *
+ * Copyright EPFL.
+ *
+ * Licensed under Apache License 2.0
+ * (https://www.apache.org/licenses/LICENSE-2.0).
+ *
+ * See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.
+ */
+
+// Contains original work for ScalaNative which falls under the Scala Native
+// license.
+
+package java.util
+
+/// ScalaNative Porting Note:
+///
+///     * Ported, with thanks & gratitude, from ScalaJS ArrayDeque.scala
+///       commit 9DC4D5b, dated 2018-10-12.
+///
+///     * ScalaNative ArrayList is the inner type, rather than js.Array.
+///
+///     * equals() & hashcode() are not implemented so comparing ArrayDeque
+///       instances ad1 == ad2 is bit of a pain. It gives object equality
+///       not the more useful contents equality. Someday...
+///           https://alvinalexander.com/scala/ \
+///           how-to-define-equals-hashcode-methods-in-scala-object-equality
+///
+///     * spliterator() is implemented but commented out because the
+///       ArrayList.spliterator it delegates to is not yet implemented.
+///
+///     * The order of method declarations is not alphabetical to reduce
+///       churn versus ScalaJs original.
+
+class ArrayDeque[E] private (private var inner: ArrayList[E])
+    extends AbstractCollection[E]
+    with Deque[E]
+    with Cloneable
+    with Serializable {
+  self =>
+
+  private var status = 0
+
+  def this() =
+    this(new ArrayList[E](16))
+
+  def this(initialCapacity: Int) = {
+    // This is the JVM behavior for negative initialCapacity.
+    this(new ArrayList[E](Math.max(0, initialCapacity)))
+  }
+
+  def this(c: Collection[_ <: E]) = {
+    this(c.size)
+    addAll(c)
+  }
+
+  override def add(e: E): Boolean = {
+    offerLast(e)
+    true
+  }
+
+  def addFirst(e: E): Unit =
+    offerFirst(e)
+
+  def addLast(e: E): Unit =
+    offerLast(e)
+
+  // shallow-copy
+  override def clone(): ArrayDeque[E] =
+    new ArrayDeque[E](inner.clone.asInstanceOf[ArrayList[E]])
+
+  def offerFirst(e: E): Boolean = {
+    if (e == null) {
+      throw new NullPointerException()
+    } else {
+      inner.add(0, e)
+      status += 1
+      true
+    }
+  }
+
+  def offerLast(e: E): Boolean = {
+    if (e == null) {
+      throw new NullPointerException()
+    } else {
+      inner.add(e)
+      status += 1
+      true
+    }
+  }
+
+  def removeFirst(): E = {
+    if (inner.isEmpty)
+      throw new NoSuchElementException()
+    else
+      pollFirst()
+  }
+
+  def removeLast(): E = {
+    if (inner.isEmpty)
+      throw new NoSuchElementException()
+    else
+      pollLast()
+  }
+
+  def pollFirst(): E = {
+    if (inner.isEmpty) null.asInstanceOf[E]
+    else {
+      val res = inner.remove(0)
+      status += 1
+      res
+    }
+  }
+
+  def pollLast(): E = {
+    if (inner.isEmpty) null.asInstanceOf[E]
+    else inner.remove(inner.size - 1)
```

Comment:
The `else` branch should also increase the `status`, like in `pollFirst()`.

Reply:
Done. oops!